File: tcp-ip/ch04/go/client/echo_client.go
package main

import (
	"fmt"
	"net"
	"os"
)

const BUF_SIZE = 30

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <serverAddr>:<serverPort>\n", os.Args[0])
		os.Exit(1)
	}
	serverAddr := os.Args[1]
	//* 客户端调用 net.Dial 函数，向服务器发送连接请求
	conn, err := net.Dial("tcp", serverAddr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[ERROR] %s\n", err.Error())
		os.Exit(1)
	}
	send := make([]byte, BUF_SIZE)
	recv := make([]byte, BUF_SIZE)
	for {
		fmt.Print("Input: ")
		fmt.Scanf("%s\n", &send)
		_ /* writeBytes */, err := conn.Write(send)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[ERROR] %s\n", err.Error())
			break
		}
		readBytes, err := conn.Read(recv)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[ERROR] %s\n", err.Error())
			break
		}
		fmt.Printf("Echo from server %s\n", recv[:readBytes])
	}
	conn.Close()
}

File: tcp-ip/ch04/echo_client.cc
//
// Created by Tiancheng on 2024/8/18.
//

#include <arpa/inet.h>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <sys/socket.h>
#include <unistd.h>

#define BUF_SIZE 30

int main(int argc, char *argv[]) {
    if (argc != 3) {
        printf("Usage: %s <serverAddr> <serverPort>\n", argv[0]);
        exit(1);
    }
    //* 调用 socket 函数，创建 socket 套接字
    //* 如果继续调用 bind, listen 函数，将成为服务器套接字
    //* 如果继续调用 connect 函数，将成为客户端套接字
    int clientSocketFd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (clientSocketFd == -1) {
        printf("Error created socket\n");
    }

    sockaddr_in serverAddr{};
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_addr.s_addr = inet_addr(argv[1]);
    serverAddr.sin_port = htons(atoi(argv[2]));

    //* 客户端调用 connect 函数，向服务器发送连接请求
    if (connect(clientSocketFd, (sockaddr *)&serverAddr, sizeof(serverAddr)) ==
        -1) {
        printf("Error connected to server\n");
        exit(1);
    }

    char send[BUF_SIZE];
    char recv[BUF_SIZE];
    while (true) {
        fputs("Input: ", stdout);
        fgets(send, BUF_SIZE, stdin);
        // send[strlen(send) - 1] = '\0'; // 将 \n 替换为 \0
        write(clientSocketFd, send, strlen(send));
        // memset(recv, '\0', BUF_SIZE); //! 将 recv 中所有元素置为 '\0'
        int readBytes = read(clientSocketFd, recv, BUF_SIZE);
        if (readBytes == -1) {
            break;
        }
        recv[readBytes] = '\0'; //! 隔离脏数据
        printf("Echo from server %s\n", recv);
    }
    close(clientSocketFd);
    return 0;
}
